### PR TITLE
docs: Describe emerald-dev docker image for local development

### PR DESCRIPTION
**Pull request by matevz**
This PR:
- Describes the usage of `emerald-dev` docker image and local network configuration in the *Writing dApps on Emerald* tutorial.
- Moves the ethereum-oasis mnemonics derivation exercise to FAQ
- Adds Remix screenshot showing the "Solidity compiler" options
- Removes double paragraphs at the end from the gitbook import leftovers
- Fixes numerous grammar errors, shortens lines, moves links to the end of the chapter and uses short references instead

### CODE REVIEW

**Comment by kostko** [comment]
Task linked: [CU-21zy6am Hands-on Workshop - how to build & deploy an Ethereum dApp on Emerald ParaTime](https://app.clickup.com/t/21zy6am)

**Review by tjanez — COMMENTED** [review]
Add removal to the [docs: Move ethereum VS oasis wallet task from dApp development to FAQ](https://github.com/oasisprotocol/docs/pull/15/commits/249491427c0ae8eb7cadc3d937373e99f5e4ed3c) commit.

File: docs/general/manage-tokens/faq.md [hunk 1]
```diff
@@ -49,6 +49,25 @@ Ensure account address shown on device's screen matches the outputted address.
...
+
+Here's a task for you:
+
+1. Visit [https://iancoleman.io/bip39/](https://iancoleman.io/bip39/) to generate a BIP39 mnemonic.
+2. Select ETH token and copy the hex-encoded private key of the first derived account, for example `0xab2c4f3bc70d40f12f6030750fe452448b5464114cbfc46704edeef2cd06da74.`
+3. Import the Ethereum-compatible account with the private key obtained above to your Oasis Wallet Browser Extension.
+4. Notice the Ethereum address of the account, for example `0x58c72Eb040Dd0DF10882aA87a39851c21Ae5F331.`
+5. Now in the Account management screen, select this account and click on the "Export private key" button. Confirm the risk warning.
+6. You will notice the private key of the Ethereum-compatible account, the hex-encoded address and the very same address encoded in the Oasis bech32 format, in our case`oasis1qpaj6hznytpvyvalmsdg8vw5fzlpftpw7g7ku0h0.`
```
Comment (tjanez) [inline docs/general/manage-tokens/faq.md]: bech32 -> Bech 32

File: docs/general/manage-tokens/faq.md [hunk 2]
```diff
@@ -49,6 +49,25 @@ Ensure account address shown on device's screen matches the outputted address.
...
 
+### Are Ethereum and Oasis wallets that different? I can use the same mnemonics with both, right?
+
+Yes, both Oasis and Ethereum wallets make use of the mnemonics as defined in [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) and they even use the same wordlist to derive the keypairs for your wallet. However, they use a different **signature scheme and a derivation path**, so the addresses and the private keys are incompatible.
+
+Here's a task for you:
+
+1. Visit [https://iancoleman.io/bip39/](https://iancoleman.io/bip39/) to generate a BIP39 mnemonic.
+2. Select ETH token and copy the hex-encoded private key of the first derived account, for example `0xab2c4f3bc70d40f12f6030750fe452448b5464114cbfc46704edeef2cd06da74.`
```
Comment (tjanez) [inline docs/general/manage-tokens/faq.md]: Please, move the `.` outside all code-styled addresses.

E.g.:
``` `0xab2c4f3bc70d40f12f6030750fe452448b5464114cbfc46704edeef2cd06da74.` ``` ->
``` `0xab2c4f3bc70d40f12f6030750fe452448b5464114cbfc46704edeef2cd06da74`. ```

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 3]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
 
 # Writing dApps on Emerald
 
-This tutorial will show you how to set up a dApp development environment using Hardhat and Truffle deployment tools or the Remix IDE. Oasis Emerald exposes an EVM-compatible interface, so writing dApps isn't much different compared to the original Ethereum network. However, there are still some important differences which this guide will walk you through.
+This tutorial will show you how to set up dApp development environment for
+Emerald. We will walk you through the Hardhat and Truffle configuration and -
+for those who prefer a simpler web-only interface - the Remix IDE. Oasis
+Emerald exposes an EVM-compatible interface, so writing dApps isn't much
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: interface, so -> omit comma

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 4]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+different compared to the original Ethereum network!
+
+## Oasis Consensus Layer and Emerald ParaTime
+
+Oasis network consists of the Consensus layer and a number of Layer 2 chains
+called the ParaTimes (to learn more, check the [Oasis Network Overview]
+[overview] chapter). Emerald is a ParaTime which implements the Ethereum
+Virtual Machine.
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Ethereum Virtual Machine -> Ethereum Virtual Machine (EVM)

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 5]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+
+The minimum and also expected block time in Emerald is **6 seconds**. Any
+Emerald transaction will require at least this amount of time to be executed.
+
+The native Oasis addresses are bech32-encoded (e.g. `oasis1qpupfu7e2n6pkezeaw0yhj8mcem8anj64ytrayne`)
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Please, change everywhere:
bech32 -> Bech32

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 6]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+while Emerald supports both the bech32-encoded and the Ethereum-compatible hex-
+encoded addresses (e.g. `0x90adE3B7065fa715c7a150313877dF1d33e777D5`). The
+underlying algorithm for signing the transactions is [ed25519] on the Consensus
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Rename ed25519 -> Ed25519 everywhere

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 7]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+called the ParaTimes (to learn more, check the [Oasis Network Overview]
+[overview] chapter). Emerald is a ParaTime which implements the Ethereum
+Virtual Machine.
+
+The minimum and also expected block time in Emerald is **6 seconds**. Any
+Emerald transaction will require at least this amount of time to be executed.
+
+The native Oasis addresses are bech32-encoded (e.g. `oasis1qpupfu7e2n6pkezeaw0yhj8mcem8anj64ytrayne`)
+while Emerald supports both the bech32-encoded and the Ethereum-compatible hex-
+encoded addresses (e.g. `0x90adE3B7065fa715c7a150313877dF1d33e777D5`). The
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Put `hex-encoded` on a single line so it doesn't create a space in the final rendered HTML.

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 8]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
-This tutorial will show you how to set up a dApp development environment using Hardhat and Truffle deployment tools or the Remix IDE. Oasis Emerald exposes an EVM-compatible interface, so writing dApps isn't much different compared to the original Ethereum network. However, there are still some important differences which this guide will walk you through.
+This tutorial will show you how to set up dApp development environment for
+Emerald. We will walk you through the Hardhat and Truffle configuration and -
+for those who prefer a simpler web-only interface - the Remix IDE. Oasis
+Emerald exposes an EVM-compatible interface, so writing dApps isn't much
+different compared to the original Ethereum network!
+
+## Oasis Consensus Layer and Emerald ParaTime
+
+Oasis network consists of the Consensus layer and a number of Layer 2 chains
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Change everywhere in the doc:
Oasis network -> Oasis Network

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 9]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+state, a different set of validators and ParaTimes. The state of the Mainnet is
+considered immutable for indefinite time, while the data on the Testnet can be
+subject to wipe in the future.
 
-Emerald is deployed similarly: the [Emerald Mainnet](./README.mdx#mainnet) is deployed on the Oasis Mainnet network, and the [Emerald Testnet](./README.mdx#testnet) on the Oasis Testnet network. The Emerald state on the Mainnet is stable. Testnet, apart from running the unstable version of the code and being prone to bugs, can have the state deliberately wiped either on the Emerald ParaTime layer or on the Oasis Testnet network level, so you should **never deploy a production service on the Testnet**!
+The Emerald ParaTime is deployed similarly: the [Emerald Mainnet] is deployed on the Oasis Mainnet
+network while the [Emerald Testnet] on the Oasis Testnet network. The Emerald state on the Mainnet
+is stable. Testnet, apart from running the unstable version of the code and being prone to bugs, can
+have the state deliberately wiped either on the Emerald ParaTime layer or on the Oasis Testnet
+network level, so you should **never deploy a production service on the Testnet**!
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Move the "you should **never deploy a production service on the Testnet**!" part to a danger tool tip.

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 10]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+[Testnet]: ../../foundation/testnet/README.md
+[Emerald Mainnet]: ./README.mdx#mainnet
+[Emerald Testnet]: ./README.mdx#testnet
 
-On the other hand, the Oasis addresses are bech32-encoded (e.g. `oasis1qpupfu7e2n6pkezeaw0yhj8mcem8anj64ytrayne`) while Emerald uses the Ethereum-compatible hex-encoded addresses (e.g. `0x90adE3B7065fa715c7a150313877dF1d33e777D5`). This is not just a stylistic change - the underlying algorithm for signing the transactions is [ed25519](https://en.wikipedia.org/wiki/EdDSA#Ed25519) on Oasis and [ECDSA](https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm) on Emerald (the same as on the Ethereum network). For this reason we cannot use Oasis accounts directly in Emerald, but we need to use a special mechanism described in the [How to transfer ROSE into Emerald ParaTime](../../manage-tokens/how-to-transfer-rose-into-emerald-paratime.mdx) chapter. To deploy your smart contract on the Emerald Mainnet, you will have to perform this step.
+## Running Private Oasis Network Locally
+
+For convenient development and testing of your dApps the Oasis team prepared a
+docker image which brings you a complete Oasis stack to your desktop. This
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Replace everywhere:
docker -> Docker

Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: `prepared a docker image` -> `prepared the [oasisprotocol/emerald-dev](link to Docker Hub) Docker image`

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 11]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+network is isolated from the Mainnet or Testnet and consists of:
+- one Oasis validator node with 1 second block time,
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: one -> single

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 12]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+- three Oasis compute nodes running Emerald ParaTime,
+- Emerald Web3 gateway with transaction indexer,
+- a helper script which populates initial test accounts for you.
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: a helper -> helper

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 13]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+
+To run the image, type:
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: type -> execute

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 14]
```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
...
+Base HD Path:   m/44'/60'/0'/0/%d
+
+WARNING: Emerald is running in ephemeral mode. The chain state will be lost after you quit!
+
+Listening on http://localhost:8545 and ws://localhost:8546
+```
+
+Those familiar with local dApp environments will find the output above similar
+to `geth --dev` or `ganache-cli` commands or the `geth-dev-assistant` npm
+package. `emerald-dev` will spin up a private Oasis network locally, generate
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Could you replace
``` `emerald-dev` ``` with ``` The [oasisprotocol/emerald-dev] ``` or ``` The [emerald-dev] ```

And similarly, use ``` [oasisprotocol/emerald-dev] ``` or ``` [emerald-dev] ``` throughout the document.

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md (hunk 10 above)
Comment (tjanez) [reply docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Also move the link to the bottom so you can reference it multiple times in the text.

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md [hunk 15]
```diff
@@ -341,56 +549,99 @@ Summary
...
-Emerald validators, similar to Ethereum ones, order the execution of transactions by gas price. When deploying a contract and the deployment times out, first wait another few rounds to make sure that the contract will not be deployed eventually.
+[Remix]: https://remix.ethereum.org
+[metamask]: ../../manage-tokens/how-to-transfer-rose-into-emerald-paratime.mdx#verifying-rose-balance-on-emerald-paratime
+[slack-devsupport]: https://oasiscommunity.slack.com/messages/dev-support
 
-Emerald validators, similar to Ethereum ones, order the execution of transactions by gas price. When deploying a contract and the deployment times out, first wait another few rounds to make sure that the contract will not be deployed eventually.
+## Troubleshooting
 
-Next, check that your `gasPrice` **is at least 10 nROSE** which is a minimum required gas price on Emerald. This value should already be propagated automatically by the web3 endpoint, but your deployment configuration might have ignored it.
+### Deployment of my contract timed out on TestNet or MainNet
```
Comment (tjanez) [inline docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: TestNet -> Testnet
MainNet -> Mainnet

**Comment by tjanez** [comment]
> * Fixes numerous grammar errors, shortens lines, moves links to the end of the chapter and uses short references instead

Thanks for this one, it is a lot more pleasant to edit and review :slightly_smiling_face:

File: docs/general/manage-tokens/faq.md (hunk 1 above)
Comment (tjanez) [reply docs/general/manage-tokens/faq.md]: Sorry, I didn't mean "Bech 32" but "Bech32".

Could you update it again so it is without the space in the middle?

File: docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md (hunk 5 above)
Comment (tjanez) [reply docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md]: Here I had "Bech32" without space :slightly_smiling_face:

**Review by tjanez — APPROVED** [review]
Just one nit with Bech32, otherwise looks good!

File: docs/general/manage-tokens/faq.md (hunk 1 above)
Comment (matevz) [reply docs/general/manage-tokens/faq.md]: Fixed.

**Pull request merged**